### DISCUSSION TITLE
How to get the selected text in textArea using react-hook-form Apis (Controlled and Uncontrolled)

### DISCUSSION THREAD

**Discussion by marzoukali** [react-hook-form/react-hook-form#9511] [Q&A]
Hello

I am trying to get the selected text in text area when I click on button.

I searched a lot and had a look at the API documentation but it seems not straight forward.

I want in Controlled or Uncontrolled textArea to be able to get the selected text inside the textArea  when you click on button.

Example of controlled TextField that I want to get it's highlighted or selected text. Below I have controlled TextField inside the Form. I want to get the selected text when I click on any button.

```
 const {
    control,
    handleSubmit,
    reset,
    formState: { errors, isValid },
    setFocus
  } = useForm({
    defaultValues,
    mode: 'onChange',
    resolver: yupResolver(schema)
  })

 

  const onSubmit = () => {
    toast.success('Form Submitted')
  }

  return (
    <Card>
      <CardContent>
        <form onSubmit={handleSubmit(onSubmit)}>
          <Grid container spacing={5}>
            <Grid item xs={12}>
              <FormControl fullWidth>
                <Controller
                  name='courseTitle'
                  control={control}
                  rules={{ required: true }}
                  render={({ field: { value, onChange } }) => (
                    <TextField
                      value={value}
                      label='Course Title'
                      onChange={onChange}
                      placeholder='Course Title'
                      error={Boolean(errors.courseTitle)}
                      aria-describedby='validation-schema-course-title'
                    />
                  )}
                />
                {errors.courseTitle && (
                  <FormHelperText sx={{ color: 'error.main' }} id='validation-schema-course-title'>
                    {errors.courseTitle.message}
                  </FormHelperText>
                )}
              </FormControl>
            </Grid>

            <Grid item xs={12}>
              <Button size='large' type='submit' variant='contained' disabled={!isValid}>
                Save
              </Button>
            </Grid>
          </Grid>
        </form>
      </CardContent>
    </Card>
  )
}

export default CourseEditForm
```

**Comment by bluebill1049** [comment]
> I want in Controlled or Uncontrolled textArea to be able to get the selected text inside the textArea when you click on button.

what do you mean by this? controlled or uncontrolled is responsible for the component level, once you are using `Controller`, the component is controlled.

**Reply by marzoukali** [reply]
What I mean, that is there any way through react-hook-form that allow me to detect the selected text in textArea? 

I was able to do it this way:

Firstly add ref to the textArea this way:
```
  const { ref } = register('courseOverview')
  const courseOverviewRef = useRef<HTMLTextAreaElement | null>(null)

    <textArea
          {...register('courseOverview', { required: true })}
          placeholder='course Overview'
          autoComplete='off'
          autoCorrect='off'
          autoCapitalize='off'
          spellCheck={false}
          ref={e => {
            ref(e)
            courseOverviewRef.current = e
          }}
        />
```

Then was able to use this ref to get the selected text.
```
    const textVal = courseOverviewRef.current!
    const cursorStart = textVal!.selectionStart
    const cursorEnd = textVal!.selectionEnd
    const highlighted = textVal.value.substring(cursorStart, cursorEnd)
```

I was wondering if this is the way or there is easiest way to do it through react-hook-form? Also if I am dealing with controlled component like the mui textField I can't use ref so how can I do it in this case?

**Reply by Phibee** [reply]
@marzoukali, In my opinion. I don't think it needs to be handle by react hook form. It needs to be handle outside the react hook form. But In relation to your concern on MUi Textfield, I provided sample code. You need to use `inputRef` instead of `ref` coz it's wrapped with div element.

Please see code below for the sample:

https://codesandbox.io/s/nervous-engelbart-hpobe7
